File: zig/src/inline/strikethroughRule.zig
const std = @import("std");
const InlineParserState = @import("../types/InlineParserState.zig").InlineParserState;
const InlineRule = @import("../types/InlineRule.zig").InlineRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const testTagMarks = @import("tagMarksRule.zig").testTagMarks;

pub fn testStrikethrough(state: *InlineParserState, parent: *MarkdownNode) bool {
    if (!state.isEscaped and state.i < state.src.len and state.src[state.i] == '~') {
        return testTagMarks("strikethrough", '~', state, parent, strikethroughRule.precedence.?);
    }
    return false;
}

pub const strikethroughRule = InlineRule{
    .name = "strikethrough",
    .@"test" = testStrikethrough,
    .precedence = 5,
};

File: web/src/inline/strikethroughRule.ts
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import { TILDE_CODE } from "../utils/charCodes";
import testTagMarks from "./tagMarksRule";

const rule: InlineRule = {
	name: "strikethrough",
	test: testStrikethrough,
	precedence: 5,
};
export default rule;

/**
 * "Strikethrough text is any text wrapped in a matching pair of one or two
 * tildes (~).
 */
function testStrikethrough(state: InlineParserState, parent: MarkdownNode): boolean {
	if (!state.isEscaped && state.src.charCodeAt(state.i) === TILDE_CODE) {
		return testTagMarks(rule.name, "~", state, parent, rule.precedence!);
	}
	return false;
}
